File: sphere.py
import math

from color import Color
from material import Material
from point import Point


class Sphere:
    def __init__(self, center: Point, radius: float, material: Material):
        self.center = center
        self.radius = radius
        self.material = material

    def intersects(self, ray):
        """Checks if ray intersects sphere. Returns distance or None"""
        sphere_to_ray = ray.origin - self.center
        b = 2 * ray.direction.dot(sphere_to_ray)
        c = sphere_to_ray.dot(sphere_to_ray) - self.radius ** 2
        discriminant = b ** 2 - 4 * c
        if discriminant >= 0:
            distance = (-b - math.sqrt(discriminant)) / 2
            if distance > 0:
                return distance
        return None

    def normal(self, surface_point: Point):
        """Returns surface normal to the point on sphere surface"""
        return (surface_point - self.center).normalize()

File: rust-tracer/src/sphere.rs
use serde::{Serialize, Deserialize};
use crate::{Point, Ray, Material, Vector3};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Sphere {
    pub center: Point,
    pub radius: f64,
    pub material: Material,
}

impl Sphere {
    pub fn new(center: Point, radius: f64, material: Material) -> Self {
        Sphere {
            center,
            radius,
            material,
        }
    }

    pub fn intersects(&self, ray: &Ray) -> Option<f64> {
        let sphere_to_ray = ray.origin - self.center;
        let b = 2.0 * ray.direction.dot(&sphere_to_ray);
        let c = sphere_to_ray.dot(&sphere_to_ray) - self.radius * self.radius;
        let discriminant = b * b - 4.0 * c;

        if discriminant >= 0.0 {
            let distance = (-b - discriminant.sqrt()) / 2.0;
            if distance > 0.0 {
                return Some(distance);
            }
        }
        None
    }

    pub fn normal(&self, surface_point: Point) -> Vector3 {
        (surface_point - self.center).normalize()
    }
}
